### PR TITLE
DR-1130 Create infra and pipelines for CM prototype

### PR DESCRIPTION
**Pull request by mmg-nhse**
Create infrastructure and GitHub workflow pipelines for the CM prototype as the prototype's source code has been relocated to a new repository.

Prototype infrastructure consists of:
* CloudFront
* WAF
* S3

Account infrastructure consists of:
* IAM role for GitHub runner
* S3 for Terraform state lock
* DynamoDB for Terraform state lock

### CODE REVIEW

File: infrastructure/modules/s3/main.tf [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  website = var.website_map
+
+  # TODO Set up access logging bucket for CSOC
+  #logging = {
+  #  target_bucket = var.target_bucket
+  #  target_prefix = var.target_prefix
+  #}
+  versioning = {
+    enabled = true
+  }
```
Comment (Tomdango) [inline infrastructure/modules/s3/main.tf]: Tech debt for future - decide if we want versioning on asset buckets - if so, add a noncurrent lifecycle configuration

File: infrastructure/stacks/github-runner/role.tf [hunk 2]
```diff
@@ -0,0 +1,95 @@
+
+# aws_iam_role_policy = an inline policy
+# aws_iam_policy, that is a managed policy and can be re-used
+# Get the policy by name
+data "aws_iam_policy" "power_user_policy" {
+  name = "PowerUserAccess"
+}
```
Comment (Tomdango) [inline infrastructure/stacks/github-runner/role.tf:7]: Tech debt - to be restricted down in future

File: infrastructure/stacks/github-runner/role.tf [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  {
+    "Version":"2012-10-17",
+    "Statement": [
+
+        {
+            "Action": "s3:*",
+            "Effect": "Allow",
+            "Resource": "*",
+            "Sid": "VisualEditor5"
+        },
```
Comment (Tomdango) [inline infrastructure/stacks/github-runner/role.tf:38]: Tech debt - to clean up as this block makes the below redundant (+ formatting)

File: infrastructure/stacks/github-runner/role.tf [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            "ForAllValues:StringLike":{
+                "token.actions.githubusercontent.com:sub":"repo:${var.github_org}/${var.repo_name}:*",
+                "token.actions.githubusercontent.com:aud":"sts.amazonaws.com"
+              }
+          }
+        }
+      ]
+    }
+    EOF
+}
```
Comment (Tomdango) [inline infrastructure/stacks/github-runner/role.tf:93]: Future tech debt - to be part of management/separated out from application deployment

File: infrastructure/stacks/github-runner/uec-github-runner-iam-services.json [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+              "iam:DeleteRolePolicy",
+              "iam:PutRolePolicy",
+              "iam:UpdateOpenIDConnectProviderThumbprint",
+              "iam:UntagPolicy",
+              "iam:UntagRole"
+          ],
+          "Effect": "Allow",
+          "Resource": "*",
+          "Sid": "VisualEditor0"
+      }
```
Comment (Tomdango) [inline infrastructure/stacks/github-runner/uec-github-runner-iam-services.json:32]: Future tech debt - deny change access to the current role

**Review by Tomdango — APPROVED** [review]
Looks good - just a few items to raise as tech debt but not to be resolved now.

File: infrastructure/common/provider.tf [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+data "aws_caller_identity" "current" {}
+
+provider "aws" {
+  region = var.aws_region
+
+  skip_metadata_api_check     = true
+  skip_region_validation      = true
+  skip_credentials_validation = true
+
+  skip_requesting_account_id = false
```
[the remark this answers is not in the document]
Comment (mmg-nhse) [reply infrastructure/common/provider.tf:15]: Tech debt ticket raised - https://nhsd-jira.digital.nhs.uk/browse/DR-1172

File: infrastructure/modules/dynamodb/main.tf [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+  name                = "${var.table_name}${local.workspace_suffix}"
+  hash_key            = var.hash_key
+  autoscaling_enabled = var.autoscaling_enabled
+  stream_enabled      = var.stream_enabled
+  stream_view_type    = var.stream_view_type
+  attributes          = var.attributes
+
+  server_side_encryption_enabled = true
+}
```
[the remark this answers is not in the document]
Comment (mmg-nhse) [reply infrastructure/modules/dynamodb/main.tf:13]: Tech debt ticket raised - https://nhsd-jira.digital.nhs.uk/browse/DR-1173

File: infrastructure/stacks/github-runner/role.tf (hunk 4 above)
Comment (mmg-nhse) [reply infrastructure/stacks/github-runner/role.tf:93]: Tech debt ticket raised - https://nhsd-jira.digital.nhs.uk/browse/DR-1173

File: infrastructure/stacks/github-runner/role.tf (hunk 3 above)
Comment (mmg-nhse) [reply infrastructure/stacks/github-runner/role.tf:38]: Tech debt ticket raised - https://nhsd-jira.digital.nhs.uk/browse/DR-1173

File: infrastructure/stacks/github-runner/uec-github-runner-iam-services.json (hunk 5 above)
Comment (mmg-nhse) [reply infrastructure/stacks/github-runner/uec-github-runner-iam-services.json:32]: Tech debt ticket raised - https://nhsd-jira.digital.nhs.uk/browse/DR-1173

File: infrastructure/modules/s3/main.tf (hunk 1 above)
Comment (mmg-nhse) [reply infrastructure/modules/s3/main.tf]: Tech debt ticket raised - https://nhsd-jira.digital.nhs.uk/browse/DR-1175

File: infrastructure/stacks/github-runner/role.tf (hunk 2 above)
Comment (mmg-nhse) [reply infrastructure/stacks/github-runner/role.tf:7]: Tech debt ticket raised - https://nhsd-jira.digital.nhs.uk/browse/DR-1176

**Pull request merged**